### PR TITLE
filter: remove needless empty destructors and problematic const

### LINKED ISSUE #3757: Disallow copy broke MSVC/clang-cl
The change in https://github.com/gnuradio/gnuradio/pull/3619 has broken MSVC + clang-cl build for me. From clang-cl while compiling e.g. `filterbank.cc` and others:
```c
In file included from f:\gv\VC_2019\VC\Tools\MSVC\14.27.29110\include\string:11:
In file included from f:\gv\VC_2019\VC\Tools\MSVC\14.27.29110\include\xstring:14:
f:\gv\VC_2019\VC\Tools\MSVC\14.27.29110\include\xmemory(694,76): error: call to deleted constructor of
      'gr::filter::kernel::fir_filter<std::complex<float>, std::complex<float>, float>'
        ::new (const_cast<void*>(static_cast<const volatile void*>(_Ptr))) _Objty(_STD forward<_Types>(_Args)...);
                                                                           ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~
f:\gv\VC_2019\VC\Tools\MSVC\14.27.29110\include\xmemory(1508,35): note: in instantiation of function template
      specialization 'std::_Default_allocator_traits<std::allocator<gr::filter::kernel::fir_filter<std::complex<float>,
      std::complex<float>, float> > >::construct<gr::filter::kernel::fir_filter<std::complex<float>,
      std::complex<float>, float>, gr::filter::kernel::fir_filter<std::complex<float>, std::complex<float>, float> &>'
      requested here
        allocator_traits<_Alloc>::construct(_Al, _Unfancy(_Last), _STD forward<_Types>(_Vals)...);
                                  ^
...
F:/gv/dx-radio/gnuradio/gv-build/include\gnuradio/filter/fir_filter.h(41,5): note: 'fir_filter' has been explicitly
      marked deleted here
    fir_filter(const fir_filter&) = delete;
    ^
```
In my lack of C++ knowledge, I fail to understand what the problem is. It seems to work with GNU-C++ since this change has been there for almost 4 months.

But trying a:
```c
    fir_filter (const fir_filter &);
    fir_filter (fir_filter &&) = default;
    fir_filter & operator = (const fir_filter &);
    fir_filter & operator = (fir_filter &&) = default;

```

causes 6 unresolved symbols during link.

### PR DESCRIPTION
**Pull request by ThomasHabets**
You can't move from a const object, so unfortunately it's not possible
to keep constness on these.

This should fix #3757

### CODE REVIEW

**Comment by ThomasHabets** [comment]
I don't know what's up with the failing test. Since the details page takes a couple of minutes to load, and then complains about .git/object/… files being removed, I'm suspecting it's broken.

File: gr-filter/include/gnuradio/filter/mmse_interp_differentiator_ff.h [hunk 1]
```diff
@@ -37,7 +37,6 @@ class FILTER_API mmse_interp_differentiator_ff
 {
 public:
     mmse_interp_differentiator_ff();
-    ~mmse_interp_differentiator_ff();
```
Comment (ryanvolz) [inline gr-filter/include/gnuradio/filter/mmse_interp_differentiator_ff.h:40]: Missed a `const` below.

**Review by ryanvolz — CHANGES_REQUESTED** [review]
I tried this with my Windows VM using MSVC 2017. I still see some errors due to trying to use deleted copy assignment:
```
error C2280: 'gr::filter::kernel::fir_filter<gr_complex,gr_complex,float> &gr::filter::kernel::fir_filter<gr_complex,gr_complex,float>::operator =(const gr::filter::kernel::fir_filter<gr_complex,gr_complex,float> &)': attempting to reference a deleted function
..\gr-filter\lib\..\include\gnuradio/filter/fir_filter.h(35): note: see declaration of 'gr::filter::kernel::fir_filter<gr_complex,gr_complex,float>::operator ='
..\gr-filter\lib\..\include\gnuradio/filter/fir_filter.h(35): note: 'gr::filter::kernel::fir_filter<gr_complex,gr_complex,float> &gr::filter::kernel::fir_filter<gr_complex,gr_complex,float>::operator =(const gr::filter::kernel::fir_filter<gr_complex,gr_complex,float> &)': function was explicitly deleted
```

The following diff was sufficient for me to get it to compile:
```
diff --git a/gr-filter/include/gnuradio/filter/filterbank.h b/gr-filter/include/gnuradio/filter/filterbank.h
index 47235e9fb..09838b6b3 100644
--- a/gr-filter/include/gnuradio/filter/filterbank.h
+++ b/gr-filter/include/gnuradio/filter/filterbank.h
@@ -57,6 +57,8 @@ public:
      */
     filterbank(const std::vector<std::vector<float>>& taps);
 
+    filterbank(filterbank&&) = default;
+
     /*!
      * Update the filterbank's filter taps.
      *
diff --git a/gr-filter/include/gnuradio/filter/mmse_fir_interpolator_cc.h b/gr-filter/include/gnuradio/filter/mmse_fir_interpolator_cc.h
index 334496b54..b21a8b248 100644
--- a/gr-filter/include/gnuradio/filter/mmse_fir_interpolator_cc.h
+++ b/gr-filter/include/gnuradio/filter/mmse_fir_interpolator_cc.h
@@ -40,6 +40,8 @@ class FILTER_API mmse_fir_interpolator_cc
 public:
     mmse_fir_interpolator_cc();
 
+    mmse_fir_interpolator_cc(mmse_fir_interpolator_cc&&) = default;
+
     unsigned ntaps() const;
     unsigned nsteps() const;
 
diff --git a/gr-filter/include/gnuradio/filter/mmse_fir_interpolator_ff.h b/gr-filter/include/gnuradio/filter/mmse_fir_interpolator_ff.h
index 40a2cc40e..b1ae1ced4 100644
--- a/gr-filter/include/gnuradio/filter/mmse_fir_interpolator_ff.h
+++ b/gr-filter/include/gnuradio/filter/mmse_fir_interpolator_ff.h
@@ -40,6 +40,8 @@ class FILTER_API mmse_fir_interpolator_ff
 public:
     mmse_fir_interpolator_ff();
 
+    mmse_fir_interpolator_ff(mmse_fir_interpolator_ff&&) = default;
+
     unsigned ntaps() const;
     unsigned nsteps() const;
 
diff --git a/gr-filter/include/gnuradio/filter/mmse_interp_differentiator_cc.h b/gr-filter/include/gnuradio/filter/mmse_interp_differentiator_cc.h
index c99d6e764..4ce7f0f89 100644
--- a/gr-filter/include/gnuradio/filter/mmse_interp_differentiator_cc.h
+++ b/gr-filter/include/gnuradio/filter/mmse_interp_differentiator_cc.h
@@ -39,6 +39,8 @@ class FILTER_API mmse_interp_differentiator_cc
 public:
     mmse_interp_differentiator_cc();
 
+    mmse_interp_differentiator_cc(mmse_interp_differentiator_cc&&) = default;
+
     unsigned ntaps() const;
     unsigned nsteps() const;
 
diff --git a/gr-filter/include/gnuradio/filter/mmse_interp_differentiator_ff.h b/gr-filter/include/gnuradio/filter/mmse_interp_differentiator_ff.h
index f1f56bf4e..e5083c8be 100644
--- a/gr-filter/include/gnuradio/filter/mmse_interp_differentiator_ff.h
+++ b/gr-filter/include/gnuradio/filter/mmse_interp_differentiator_ff.h
@@ -38,6 +38,8 @@ class FILTER_API mmse_interp_differentiator_ff
 public:
     mmse_interp_differentiator_ff();
 
+    mmse_interp_differentiator_ff(mmse_interp_differentiator_ff&&) = default;
+
     unsigned ntaps() const;
     unsigned nsteps() const;
 
@@ -57,7 +59,7 @@ public:
     float differentiate(const float input[], float mu) const;
 
 protected:
-    const std::vector<kernel::fir_filter_fff> filters;
+    std::vector<kernel::fir_filter_fff> filters;
 };
 
 } /* namespace filter */
```

Other than the missed removal of `const` in `mmse_interp_differentiator_ff.h`, I don't know why MSVC won't declare these move constructors without setting them to default explicitly. I'd guess one of the members triggers a rule that prevents implicit declaration.

File: gr-filter/include/gnuradio/filter/mmse_interp_differentiator_ff.h (hunk 1 above)
Comment (ThomasHabets) [reply gr-filter/include/gnuradio/filter/mmse_interp_differentiator_ff.h:40]: Indeed.

Any guesses on why the move constructors have to be explicitly defaulted? [Reads to me](https://en.cppreference.com/w/cpp/language/move_constructor) under these circumstances.

I'll add your changes anyway.

**Comment by ryanvolz** [comment]
Maybe this is the culprit? https://stackoverflow.com/questions/53168836/does-visual-studio-2017-need-an-explicit-move-constructor-declaration

**Comment by ThomasHabets** [comment]
@ryanvolz so if I read that right then that would mean that there's some move constructor somewhere that's not noexcept?

I'm still not understanding it. Why does MSVC not complain about [this code](https://godbolt.org/z/shG5eE)?

**Review by ryanvolz — APPROVED** [review]

**Comment by ryanvolz** [comment]
> I'm still not understanding it. Why does MSVC not complain about [this code](https://godbolt.org/z/shG5eE)?

I don't understand it, and I can't come up with any toy examples that fail to compile either. I think this is one where we just blame MSVC and move on with what works.

**Comment by marcusmueller** [comment]
@ryanvolz uff, don't have a windows box at my disposal right now, but: this works now, right?

**Comment by ryanvolz** [comment]
> @ryanvolz uff, don't have a windows box at my disposal right now, but: this works now, right?

Yes, it works now!

**Comment by argilo** [comment]
@mormj I removed myself as a reviewer since I don't know C++ well enough to competently review this change.

**Comment by ghostop14** [comment]
I haven't built/run it on Windows either, so I'm not sure I have any constructive input.

**Comment by ThomasHabets** [comment]
So who/what is next step?

**Comment by mormj** [comment]
Is the WIP label ready to be removed?  It would be nice to get one more set of eyes on it, but logically it all makes sense, and if it fixes the windows builds ... I guess it can be merged.

**Comment by ThomasHabets** [comment]
oops, yes. after ryanvolz confirmation I'd say no longer WIP.

**Comment by mormj** [comment]
I'll go with the Windows guys confirmation on this and go ahead and merge.

**Pull request merged**